pkg/operator/staticpod/controller/prune/prune_controller.go

```diff
@@ -0,0 +1,212 @@
+package prune
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"vbom.ml/util/sortorder"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+)
+
+type PruneController struct {
+	targetNamespace, staticPodName              string
+	failedRevisionLimit, succeededRevisionLimit int
+
+	// command is the string to use for the pruning pod command
+	command []string
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+	// prunerPodImageFn returns the image name for the installer pod
+	prunerPodImageFn func() string
+
+	kubeClient    kubernetes.Interface
+	eventRecorder events.Recorder
+}
+
+const (
+	statusConfigMapName = "revision-status"
+	defaultResourceDir  = "/etc/kubernetes/static-pod-resources"
+	defaultConfigMapDir = "configmaps/revision-status/phase"
+	prunerPodName       = "revision-pruner"
+)
+
+func NewPruneController(
+	targetNamespace, staticPodName string,
+	command []string,
+	kubeClient kubernetes.Interface,
+	failedRevisionLimit, succeededRevisionLimit int,
+	eventRecorder events.Recorder,
+) *PruneController {
+	c := &PruneController{
+		targetNamespace:        targetNamespace,
+		staticPodName:          staticPodName,
+		command:                command,
+		failedRevisionLimit:    failedRevisionLimit,
+		succeededRevisionLimit: succeededRevisionLimit,
+
+		kubeClient:    kubeClient,
+		eventRecorder: eventRecorder,
+
+		queue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "PruneController"),
+		prunerPodImageFn: getPrunerPodImageFromEnv,
+	}
+
+	return c
+}
+
+func (c *PruneController) pruneRevisionHistory(resourceDir, configMapDir string) error {
+	var succeededRevisionIDs, failedRevisionIDs []string
+
+	// Walk the resource directory looking for files that match our static pod name prefix
+	err := filepath.Walk(resourceDir, func(currentPath string, info os.FileInfo, err error) error {
```

Comment:
The controller has no access to the filesystem where these defintions live.

Reply:
> The controller has no access to the filesystem where these defintions live.

seems like this would come from inspecting the configmaps via the API